Tests for NAT-Traversal and Hole-Punching

emmacasolin opened the pull request:
### Description

This PR description was copied from #357 

The architecture for our NAT busting will be completed in https://github.com/MatrixAI/js-polykey/pull/326. Once this is done, we should be able to run fully simulated tests for communication between nodes using hole-punching across NAT. This will involve writing Jest tests that create namespaces with iptables rules to simulate different kinds of NAT and creating nodes inside of these namespaces. Specifically, we are only looking at endpoint-independent and endpoint-dependent NAT mapping, since our nat busting should be able to traverse all different types of firewalls. Thus, these tests will only be simulating port-restricted cone (endpoint-independent) and symmetric (endpoint-dependent) NAT.

The test cases that need to be considered here are:
1. Node1 connect to Node2 - basic sanity test
2. Node1 behind NAT connects to Node2 - here Node1 is acting like a client and it is behind a NAT, connecting to an open Node2 that isn't behind NAT
3. Node1 connects to Node2 behind NAT - here Node1 is acting like a client, connecting to a closed Node2 that is behind a NAT
4. Node1 behind NAT connects to Node2 behind NAT - here Node1 is acting like a client and it is behind NAT, and it is connecting to Node2 which is also behind NAT

These tests should cover all possible NAT combinations and should be repeated both with and without the use of a seed node as a signalling server (as such, some tests will be expected to fail, since the seed nodes are used in our NAT busting).

Relates to #148

### Issues Fixed
* ~~Fixes #159~~
* Fixes #161
* Fixes #380 

### Tasks
- [x] 1. Create util functions for creating namespaces and interfaces, as well as for setting up various iptables rules
- [x] 2. Write tests for basic connection (no NAT, sanity test)
- [x] 3. Write tests for one node behind NAT
- [x] 4. Write tests for both nodes behind NAT
- ~~[ ] 5. Write tests using a seed node~~ To be done in a new PR
- [x] 6. Ensure tests are separated from the rest of the testing suite (i.e. are not run with `npm run test`), and make sure they check that the operating system is Linux before running
- ~~[ ] 7. Investigate no STDERR logs when running `pk agent status` or `pk agent start` in the beginning https://github.com/MatrixAI/js-polykey/pull/357#issuecomment-1064723583 could be resolved in #326~~ - resolved here: https://github.com/MatrixAI/js-polykey/pull/357#issuecomment-1064747556
- [x] 8. Investigate the usage of `unshare` instead of `ip netns` which reduces the overhead of using `sudo` and makes our tests more accessible in other platforms like CI/CD.
- ~[ ] 9. Authenticate the sender of hold-punch messages #148~ - Not relevant to this PR, we will need to review the security of the P2P system after testnet deployment

### Final checklist
* [x] Domain specific tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

joshuakarp committed:
test: NAT Traversal testing utils and tests

emmacasolin committed:
test: `describeIf` and `testIf` utilities for conditional testing

CMCDragonkai commented on tests/nat/endpointIndependentNAT.test.ts:
```diff
@@ -1,372 +1,383 @@
-});
+    });
+    test(
+      'Node1 behind EIM NAT connects to Node2',
```
Can you make your test descriptions lowercase, all of our test descriptions are lower case atm. And it should say `node 1 behind ...`.

Also what is EIM? I'm not familiar with this acronym, perhaps it should be expanded?

CMCDragonkai commented on tests/nat/noNAT.test.ts:
```diff
@@ -2,238 +2,249 @@ import os from 'os';
+    shell.which('iptables') &&
+    shell.which('nsenter') &&
+    shell.which('unshare'),
+  'no NAT',
```
I would prefer calling this `DMZ`, to indicate no NAT situations.

emmacasolin committed:
feat: better logging for service handler errors

If an error was logged out by a service handler it would previously appear as `createClientService:`, which was too vague. The logger will now state the name of the handler.

emmacasolin committed:
fix: PolykeyAgent now stops nodes domains when startup fails

emmacasolin committed:
fix: ICE protocol uses `Promise.any` instead of `Promise.all`

emmacasolin committed:
tests: fixes from testnet changes

The testnet PR brought in some changes that affect the NAT tests, so they needed to be modified slightly.
- Adding a node to the node graph pings the node by default. We want to disable this in the NAT tests so we have more control over the pings.
- Nodes add the details of any node that pings them. We can now remove some additional `nodes add` calls that were required to imitate this previously missing functionality.

emmacasolin committed:
build: upgraded `async-init`, `async-locks` and `js-logger`

emmacasolin committed:
feat: removing `edmSimple` configuration

Now that nodes add the details of a node that contacts them, we no longer need the `edmSimple` configuration to do this manually.

emmacasolin committed:
style: linting and other style fixes

General linting, using capitals for constants in NAT utils, and lowercase test descriptions

emmacasolin committed:
fix: relay node for hole punch message now re-writes proxy address

A relay node for a hole punch message was previously not modifying the proxy address in the message (which is the "return address" used to contact the source node). For nodes behind a NAT, who do not know their own public address, they rely on this overwriting so that nodes do not try to contact them on their private, inaccessible address.

tegefaulkes committed:
feat: added ability to pause `refreshBucketQueue`

This will allow us to disable the queue for testing.

emmacasolin committed:
build: package upgrades to aid testing

emmacasolin committed:
style: lintfix

emmacasolin committed:
feat: added diagrams to DMZ tests

#388

tegefaulkes committed:
fix: async bug with composing connections

the composed flag was set at the beginning of the compose function causing another function to throw an error due to an undefined property if it was called at the same time.
